### ISSUE TITLE
[AArch64][ARM] Why fpack-struct=4 has a higher priority than attribute

### ISSUE THREAD

**Issue by LukeSTM** [llvm/llvm-project#103495]
```
struct s1
{
  int __attribute__ ((aligned (8))) a;
};

struct
{
  char c;
  struct s1 m;
} v;

int
main (void)
{
  if ((int)&v.m & 7)
    abort ();
  exit (0);
}
```
compile with only one option：-fpack-struct=4

gcc could pass this testcase, but clang could not.
https://godbolt.org/z/h5T5rdYWb

The code can be simplified as follows:
```
struct s1
{
  int __attribute__ ((aligned (8))) a;
}s;
```
https://godbolt.org/z/EbGsEz6eP
![image](https://github.com/user-attachments/assets/a052017d-5d28-4ceb-862c-e400c022492c)

if compile without -fpack-struct=4, the alignment  of struct is 8, as same as attribue. But if compile with -fpack-struct=4, the alignment  of struct is 4. Why fpack-struct=4 has a higher priority than attribute? I think attribute should has a higher priority than Compilation options

**Comment by LukeSTM** [comment]
__attribute__ ((packed)) has a higher priority than fpack-struct=4

https://godbolt.org/z/j3x8P93z3
![image](https://github.com/user-attachments/assets/aa8a9955-36d7-48f0-957f-b6886970d674)

**Comment by LukeSTM** [comment]
It seem like llvm prefer a small alignment. 
If attribute ((align (4))) and fpack-struct=8, alignment = 4.
If attribute ((align (4))) and fpack-struct, alignment = 1.
https://godbolt.org/z/ddeMvzEGE
![image](https://github.com/user-attachments/assets/f69822ba-e4dc-4a3f-9f3d-ca71fc86ba59)

**Comment by hstk30-hw** [comment]
It's this an expected behavior. CC @AaronBallman @zygoloid

**Comment by LukeSTM** [comment]
![image](https://github.com/user-attachments/assets/170a43a7-2774-41b9-afe0-8e93cfe09888)

**Comment by LukeSTM** [comment]
I understand that the alignment rules of the structures of gcc and llvm should be the same, but the default alignment rules are different.
gcc arm struct alignment default 4byte，and could not less than 4byte。llvm does not have this restriction.

**Comment by AaronBallman** [comment]
This isn't ARM-specific, you get the same behavior with x86-64, for example.

This has been Clang's behavior for at least 15 years, see 072094407d926cb9253171e022912c92a6498ae5. I wonder how much ABI breakage there would be from changing behavior now. That said, I think the attribute *should* override the command line flag because the usual user expectation is that attributes take precedence over pragmas which take precedence over command line flags. The closer you are to the source, the higher priority the marking should be.

CC @dwblaikie @erichkeane @rjmccall for ideas and opinions

**Comment by erichkeane** [comment]
> This isn't ARM-specific, you get the same behavior with x86-64, for example.
> 
> This has been Clang's behavior for at least 15 years, see [0720944](https://github.com/llvm/llvm-project/commit/072094407d926cb9253171e022912c92a6498ae5). I wonder how much ABI breakage there would be from changing behavior now. That said, I think the attribute _should_ override the command line flag because the usual user expectation is that attributes take precedence over pragmas which take precedence over command line flags. The closer you are to the source, the higher priority the marking should be.
> 
> CC @dwblaikie @erichkeane @rjmccall for ideas and opinions

I concur as far as that.  The ABI break would be unfortunate/pretty annoying.  At minimum we'd need to do the `ClangABI` tag change for this, but I'd probably want us to have some sort of warning diagnostic of "this has changed, use ClangABI=19 if you care".

**Comment by rjmccall** [comment]
It's actually less of a serious ABI issue that you'd probably think because `-fpack-struct` is unusable in most scenarios — any command line flag that changes struct layout is inherently wildly ABI-breaking.

**Comment by AaronBallman** [comment]
> It's actually less of a serious ABI issue that you'd probably think because `-fpack-struct` is unusable in most scenarios — any command line flag that changes struct layout is inherently wildly ABI-breaking.

Oh that's a good point... that flag would change packing for *every* structure, including ones pulled in via system headers.

**Comment by rjmccall** [comment]
> > It's actually less of a serious ABI issue that you'd probably think because `-fpack-struct` is unusable in most scenarios — any command line flag that changes struct layout is inherently wildly ABI-breaking.
> 
> Oh that's a good point... that flag would change packing for _every_ structure, including ones pulled in via system headers.

Right.  There's a pragma that's less global, but I'm pretty sure that turns into the attribute rather than this global setting.

**Comment by erichkeane** [comment]
Well, that still makes an an ABI break, but at least a small one (only those who use that `-fpack-struct` and use this attribute elsewhere).  So I don't have heartburn over that, but perhaps a warning if those two don't match is still a reasonable idea (plus, I think incrementing the ABI flag is a good idea anyway).

**Comment by dwblaikie** [comment]
Usually fixing to be compatible with gcc is what we do at tot - folks who want clang abi compatibility with itself would use the flag for a specific clang abi version (& folks have platforms that bake it in to a specific clang abi version of that's what they want)

Seems niche enough (& the point about this being pretty hard to use correctly anyway - probably fairly narrowly scoped usage) that we could/should just fix this at clang/tot with the usual back compat provided through -fclang-abi-version or whatever the flag is called.

**Comment by rjmccall** [comment]
Yeah, I agree with the suggestion to fix this with the usual `-fclang-abi-version` treatment.

**Comment by LukeSTM** [comment]
So，do we have any develop plan to add -fclang-abi-version？

**Comment by dwblaikie** [comment]
> So，do we have any develop plan to add -fclang-abi-version？

I think it's a "patches welcome" sort of situation, probably? If you wanted to work on a patch that makes the ABI change/break, but conditionally/only under more recent `-fclang-abi-version` numbers - send it out, and we'll review it.
